service/openapi.yaml

```diff
@@ -0,0 +1,241 @@
+openapi: 3.1.0
+info:
+  version: 0.1.0
+  title: OpenFeature Remote Evaluation Protocol (OFREP)
+  description: OFREP define the protocol for remote flag evaluations
+  contact:
+    url: https://github.com/open-feature/protocol
+  license:
+    identifier: Apache-2.0
+    name: Apache 2.0
+security:
+  - ApiKeyAuth: [ ]
+  - BearerAuth: [ ]
+paths:
+  /ofrep/v1/evaluate/flags/{key}:
+    post:
+      description: OFREP single flag evaluation request
+      parameters:
+        - name: key
+          in: path
+          required: true
+          schema:
+            type: string
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/evaluationRequest'
+      responses:
+        '200':
+          description: OFREP successful evaluation response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/evaluationSuccess'
+        '400':
```

Comment:
Shouldn't there be a 404 response for a key that isn't found?

Reply:
Seeing that we do have an error code for `"errorCode": "FLAG_NOT_FOUND",` 

Or is the intention that we always return the default value with `"reason": "DEFAULT"`, its a bit ambiguous how providers should implement this.

Reply:
Just saw the discussion here that a 404 should be captured by a `400` error: https://github.com/open-feature/protocol/pull/2#issuecomment-1965120023

I disagree that we we should be capturing `404` errors with a more generic `400` error, we should not be trying to change standard HTTP error code behaviour, for most API consumers it will be very un-natural that a not-found error is a `400` not a `404`.

Reply:
We already have HTTP 400 to communicate back bad flag evaluation responses mapping to error codes [1] and this is why I removed the dedicated 404 response. And if the provider receives a 400 (or any other error), the evaluation should be `ERROR`.

Besides, in my view flag evaluation request is not necessarily a request to obtain a resource but rather an execution of an operation. So a 404 should mean `missing operation` which translates to a missing flag evaluation endpoint which we will have to handle in the provider implementation. 

 (cc - @lukas-reining  @thomaspoignant appreciate opinion on this )

[1] - https://openfeature.dev/specification/types#error-code

Reply:
I have no strong opinion on this one.
I found the 404 handy since it is explicit from an API perspective.

But I can live without it also.

Reply:
A 404 makes sense when attempting to evaluate an individual flag.

Reply:
> I found the 404 handy since it is explicit from an API perspective.

Staying close to specific HTTP semantic generally feels like a good thing, so going for 404 for single evaluations seems good, on the other hand one might argue that all errors are expressed through the 400 with an error code in our case, which is still HTTP compliant and might give us a simpler interface. So to me, parsing 400 errors would always be necessary and 404 would not give us any additional information. 
Yes it might give us a shortcut to not have to look into the response body @thomaspoignant, still I would expect a `evaluationFailure` response here.

From the [RFC](https://httpwg.org/specs/rfc9110.html#status.401):
> A 404 response is heuristically cacheable; i.e., unless otherwise indicated by the method definition or explicit cache controls (see [Section 4.2.2](https://httpwg.org/specs/rfc9111.html#heuristic.freshness) of [[CACHING]](https://httpwg.org/specs/rfc9110.html#CACHING)).

I am not sure if this might be important to keep in mind. Maybe we do or do not want this to be possibly done by middle boxes like proxys? 

> So a 404 should could mean missing operation which translates to a missing flag evaluation endpoint which we will have to handle in the provider implementation.

By the specification a 404 for `/ofrep/v1/evaluate/flags/{key}` would always mean that the flag is not found if the service is OFREP compliant so I am not sure if that is really the case here @Kavindu-Dodan.


Generally I think both 400 and 404 would be okay. I am leaning towards a 404 with the same `evaluationFailure` response as for the other 400s.

Reply:
> am leaning towards a 404 with the same evaluationFailure response as for the other 400s.

Agreed.

Reply:
Thank you @jonathannorris for starting this conversation. And based on feedback from @thomaspoignant @beeme1mr @lukas-reining & @toddbaert  I will add a dedicated 404 response with error code set to `FLAG_NOT_FOUND `